### PR TITLE
New CSV Export

### PR DESCRIPTION
**Pull request by llvasconcellos**
The default behavior of the new CSV Export is to merge encounters that are near each other by 30 minutes.
Each encounter within 30 minutes will be merged to the previous one overlapping the old value with the most recent one.
The interval can be customized passing a value in the url like:
http://localhost:9000/openmrs/moduleServlet/buendiadata?interval=120

### CODE REVIEW

File: openmrs/omod/src/main/java/org/openmrs/projectbuendia/servlet/DataExportServlet.java [hunk 1]
```diff
@@ -89,11 +90,49 @@
     private static final int COLUMNS_PER_OBS = 3;
     private static final ClientConceptNamer NAMER = new ClientConceptNamer(Locale.ENGLISH);
 
+    public static final int DEFAULT_INTERVAL = 30;
+
+    private VisitObsValue.ObsValueVisitor stringVisitor;
+    private void defineStringVisitor() {
+        stringVisitor = new VisitObsValue.ObsValueVisitor<String>() {
```
Comment (capnfabs) [inline openmrs/omod/src/main/java/org/openmrs/projectbuendia/servlet/DataExportServlet.java]: Style note: generally, I prefer to put implementations like this at the bottom of the class definition, and to use
```java
private final VisitObsValue.ObsValueVisitor<String> stringVisitor = new VisitObsValue.ObsValueVisitor() {
//...
};
```

Then you don't have to have the separate method definition. You could even make `stringVisitor` static and then put the name in all caps, e.g. `STRING_VISITOR`.

File: openmrs/omod/src/main/java/org/openmrs/projectbuendia/servlet/DataExportServlet.java [hunk 2]
```diff
@@ -89,11 +90,49 @@
     private static final int COLUMNS_PER_OBS = 3;
     private static final ClientConceptNamer NAMER = new ClientConceptNamer(Locale.ENGLISH);
 
+    public static final int DEFAULT_INTERVAL = 30;
```
Comment (capnfabs) [inline openmrs/omod/src/main/java/org/openmrs/projectbuendia/servlet/DataExportServlet.java]: please always include units in the name for time-related constants, e.g. `DEFAULT_INTERVAL_MINS`.

File: openmrs/omod/src/main/java/org/openmrs/projectbuendia/servlet/DataExportServlet.java [hunk 3]
```diff
@@ -159,86 +214,25 @@ protected void doGet(HttpServletRequest request, HttpServletResponse response) t
...
+                            if ((value != null) && (!value.isEmpty())) {
                                 values[valueColumn] = value;
                                 values[valueColumn + 1] = value;
                                 values[valueColumn + 2] = value;
-                                return null;
                             }
-                        });
+                        }
+                    }
+                    if (i == (encounters.size() - 1)){
```
Comment (capnfabs) [inline openmrs/omod/src/main/java/org/openmrs/projectbuendia/servlet/DataExportServlet.java]: You can move the final `printer.printRecord` outside the loop and remove this conditional.

File: openmrs/omod/src/main/java/org/openmrs/projectbuendia/servlet/DataExportServlet.java [hunk 4]
```diff
@@ -89,11 +90,49 @@
...
     protected void doGet(HttpServletRequest request, HttpServletResponse response) throws
         ServletException, IOException {
-        CSVPrinter printer = new CSVPrinter(response.getWriter(), CSVFormat.EXCEL.withDelimiter
-            (','));
+
+        // Defines the interval in minutes that will be used to merge encounters.
+        int interval = DEFAULT_INTERVAL;
+        String intervalParameter = request.getParameter("interval");
+        if (intervalParameter != null) {
+            interval = Integer.valueOf(intervalParameter);
```
Comment (capnfabs) [inline openmrs/omod/src/main/java/org/openmrs/projectbuendia/servlet/DataExportServlet.java]: Maybe validate that the interval is set `> 0`?

**Comment by capnfabs** [comment]
All in all, looks good.

I've left some specific comments, but also, could you add some more comments to the logic in the method? This is quite a long function now and it would be good to have a better indication of what each step is doing.

File: openmrs/omod/src/main/java/org/openmrs/projectbuendia/servlet/DataExportServlet.java (hunk 3 above)
Comment (llvasconcellos) [inline openmrs/omod/src/main/java/org/openmrs/projectbuendia/servlet/DataExportServlet.java]: F...k you are right. 
https://www.youtube.com/watch?v=5rAOyh7YmEc

File: openmrs/omod/src/main/java/org/openmrs/projectbuendia/servlet/DataExportServlet.java (hunk 2 above)
Comment (llvasconcellos) [inline openmrs/omod/src/main/java/org/openmrs/projectbuendia/servlet/DataExportServlet.java]: Done!

File: openmrs/omod/src/main/java/org/openmrs/projectbuendia/servlet/DataExportServlet.java (hunk 4 above)
Comment (llvasconcellos) [inline openmrs/omod/src/main/java/org/openmrs/projectbuendia/servlet/DataExportServlet.java]: Done!

**Comment by capnfabs** [comment]
Nice one!

**Pull request merged**